src/admin/index.js

```diff
@@ -990,6 +994,86 @@ module.exports = ({
     })
   }
 
+  /**
+   * Delete topic records up to the selected partition offsets
+   *
+   * @param {string} topic
+   * @param {Array<SeekEntry>} partitions
+   * @return {Promise}
+   *
+   * @typedef {Object} SeekEntry
+   * @property {number} partition
+   * @property {string} offset
+   */
+  const deleteTopicRecords = async ({ topic, partitions }) => {
+    if (!topic || typeof topic !== 'string') {
+      throw new KafkaJSNonRetriableError(`Invalid topic ${topic}`)
+    }
+
+    if (!partitions || partitions.length === 0) {
+      throw new KafkaJSNonRetriableError(`Invalid partitions`)
+    }
+
+    const partitionsByBroker = cluster.findLeaderForPartitions(
+      topic,
+      partitions.map(p => p.partition)
+    )
+
+    const partitionsFound = flatten(values(partitionsByBroker))
+    const topicOffsets = await fetchTopicOffsets(topic)
+
+    partitions.forEach(({ partition, offset }) => {
+      // warn if no broker/leader found for partition
+      if (!partitionsFound.includes(partition)) {
+        logger.warn('Could not find broker/leader for the partition', {
+          topic,
+          partition,
+        })
+        return
+      }
+      const { high, low } = topicOffsets.find(p => p.partition === partition)
+      // throw in case of offset out-of-range (otherwise could fail mid-process)
+      if (parseInt(offset) > parseInt(high) || parseInt(offset) < -1) {
+        throw new KafkaJSOffsetOutOfRange(
+          'The requested offset is not within the range of offsets maintained by the server',
+          { topic, partition }
+        )
+      }
+      // warn in case of offset below low watermark
+      if (parseInt(offset) < parseInt(low)) {
+        logger.warn(
+          'The requested offset is before the earliest offset maintained on the partition - no records will be deleted from this partition',
+          {
+            topic,
+            partition,
+            offset,
+          }
+        )
+      }
+    })
+
+    const seekEntriesByBroker = entries(partitionsByBroker).reduce(
+      (obj, [nodeId, nodePartitions]) => {
+        obj[nodeId] = {
+          topic,
+          partitions: partitions.filter(p => nodePartitions.includes(p.partition)),
+        }
+        return obj
+      },
+      {}
+    )
+
+    const retrier = createRetry(retry)
+    return retrier(async () => {
+      for (const [nodeId, { topic, partitions }] of entries(seekEntriesByBroker)) {
+        const broker = await cluster.findBroker({ nodeId })
+        await broker.deleteRecords({ topics: [{ topic, partitions }] })
```

Comment:
I think we need to handle some error codes here. Specifically if we have stale metadata, so that we try to make a request to a broker that used to be the leader for a partition, but is not anymore.

[`sendMessages.js`](https://github.com/tulios/kafkajs/blob/master/src/producer/sendMessages.js) is a good example of this: https://github.com/tulios/kafkajs/blob/02bc8a3c19504d182df131914ee9af4661bf060c/src/producer/sendMessages.js#L121-L138

Reply:
Tried to implement in the latest. Had to improvise a little bit because we're not waiting on all the requests to get resolved, so it will check if the "master" `KafkaJSDeleteTopicRecordsError` error is retriable + if at least 1 of the errors was related to stale metadata. Let me know if you had something else in mind (I also moved the `staleMetadata` helper function out into the errors file - but happy to move back if it doesn't belong there)

```javascript
} catch (e) {
  if (
    e.retriable &&
    e.brokers.some(
      ({ error }) => staleMetadata(error) || error.name === 'KafkaJSMetadataNotLoaded'
    )
  ) {
    await cluster.refreshMetadata()
  }
  throw e
}
```